Fix choices ignoring the defined blank value

orlnub123 opened the pull request:
While sifting through the source code to understand the internals for a package I'm making, I stumbled upon what looked like a bug.

These lines right here:
```py
named_groups = isinstance(choices[0][1], (list, tuple))
blank_defined = not named_groups and any(choice in ('', None) for choice, __ in choices)
```
And indeed a bug it was.

The problem was that, had you defined your choices first with a group and then later on a choice that defined the blank value, the current implementation would skip right over it! So you would end up with having both the default blank value and the one you've defined, which doesn't look very pretty.

![picture](https://user-images.githubusercontent.com/30984274/37772917-39e232e8-2ded-11e8-9767-e3aa5bf98a66.png)

This pull request is my solution.

It gets rid of the group checking entirely as the documentation quotes that they are "used for organizational purposes", so it wouldn't make sense to check them, even if one did get named as either '' or None, which also addresses the issue in #23098 that those lines were put in there for in the first place.

timgraham commented:
Will you add a test to demonstrate what's changed? `tests/model_fields/tests.py` might be an appropriate location.

orlnub123 commented:
After reading the documentation again it tells you to "add a tuple to choices containing None; e.g. (None, 'Your String For Display')." So maybe it's best to ignore choices in groups when checking for a blank value?

I've updated my pull request which now ignores choices in groups with a fix that makes the blank value appear at the top and also added the requested test although I'm sure it could be improved.

timgraham commented:
I'm not sure if the implementation is correct. Why does it move the blank choice from wherever the user defined it to before all the choices? What was the behavior before 2f73b527dda6683868fac2791f7f07ccb01ea0d9? The commit where a custom blank value was first allowed is 1123f4551158b7fc65d3bd88c375a4517dcd0720.

orlnub123 commented:
Sorry for taking so long, had a busy schedule. Anyways, you're right, my updated implementation wasn't correct. I did a lot more digging and found the base of my problem.

get_choices has two problematic outcomes. Either you define a group at the top and then add a tuple for a blank value, which I had done in the example picture, or you add the tuple inside of a group. Both result in get_choices not picking up on the blank value and adding its own tuple which differs from directly defining the form field where the Select widget correctly picks up on it and adds the selected HTML attribute.

I had actually nailed it on the first try. I've reverted to the first implementation and made an adequate test for it.
